test/integration/upgrade-edge/upgrade_edge_test.go

```diff
@@ -0,0 +1,533 @@
+	t.Run(fmt.Sprintf("installing Linkerd %s control plane", linkerdBaseEdgeVersion), func(t *testing.T) {
+		args := []string{
+			"install",
+			"--controller-log-level", "debug",
+			"--set", "proxyInit.ignoreInboundPorts=1234\\,5678",
+		}
+
+		// Pipe cmd & args to `linkerd`
+		out, err := TestHelper.CmdRun(cliPath, args...)
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "'linkerd install' command failed", "'linkerd install' command failed:\n%v", err)
+		}
+
+		out, err = TestHelper.KubectlApply(out, "")
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "'kubectl apply' command failed",
+				"'kubectl apply' command failed\n%s", out)
+		}
+
+		TestHelper.WaitRollout(t, testutil.LinkerdDeployReplicasEdge)
+	})
+
+	// TestInstallViz will install the viz extension to be used by the rest of the
+	// tests in the viz suite
+	t.Run(fmt.Sprintf("installing Linkerd %s viz extension", linkerdBaseEdgeVersion), func(t *testing.T) {
+		args := []string{
+			"viz",
+			"install",
+			"--set", fmt.Sprintf("namespace=%s", TestHelper.GetVizNamespace()),
+		}
+
+		out, err := TestHelper.CmdRun(cliPath, args...)
+		if err != nil {
+			testutil.AnnotatedFatal(t, "'linkerd viz install' command failed", err)
+		}
+
+		out, err = TestHelper.KubectlApplyWithArgs(out)
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "'kubectl apply' command failed",
+				"'kubectl apply' command failed\n%s", out)
+		}
+
+		TestHelper.WaitRollout(t, testutil.LinkerdVizDeployReplicas)
+		TestHelper.AddInstalledExtension("viz")
+
+	})
+
+	// Check client and server versions are what we expect them to be
+	t.Run(fmt.Sprintf("check version is %s pre-upgrade", linkerdBaseEdgeVersion), func(t *testing.T) {
+		out, err := TestHelper.CmdRun(cliPath, "version")
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "'linkerd version' command failed", "'linkerd version' command failed\n%s", err.Error())
+		}
+
+		if !strings.Contains(out, fmt.Sprintf("Client version: %s", linkerdBaseEdgeVersion)) {
+			testutil.AnnotatedFatalf(t, "'linkerd version' command failed", "'linkerd version' command failed\nexpected client version: %s, got: %s", linkerdBaseEdgeVersion, out)
+		}
+		if !strings.Contains(out, fmt.Sprintf("Server version: %s", linkerdBaseEdgeVersion)) {
+			testutil.AnnotatedFatalf(t, "'linkerd version' command failed", "'linkerd version' command failed\nexpected server version: %s, got: %s", linkerdBaseEdgeVersion, out)
+		}
+	})
+}
+
+func TestUpgradeTestAppWorksBeforeUpgrade(t *testing.T) {
+	ctx := context.Background()
+	testAppNamespace := "upgrade-test"
+
+	// create namespace and install test app
+	if err := TestHelper.CreateDataPlaneNamespaceIfNotExists(ctx, testAppNamespace, map[string]string{}); err != nil {
+		testutil.AnnotatedFatalf(t, "Failed to create namespace", "Failed to create namespace %s: %s", testAppNamespace, err)
+	}
+
+	if _, err := TestHelper.Kubectl("", "apply", "-f", "./testdata/emoji.yaml", "-n", testAppNamespace); err != nil {
```

Comment:
Originally, emoji was injected in `bin/_test-helpers.sh` so we check the linkerd upgrade doesn't break the app. Here you can instead just annotate the emojivoto namespace for injection, and then before `TestUpgradeTestAppWorksAfterUpgrade` do a dataplane rollout.

Reply:
When creating the namespace we now add the inject annotation. Before we upgrade, we roll out the pods so they get re-injected, block until ready, and then go ahead with the rest of the test 👍🏻  great suggestion